src/Microsoft.DotNet.Build.Tasks/CleanupVSTSAgent.cs

```diff
@@ -0,0 +1,156 @@
+using Microsoft.Build.Framework;
+using Newtonsoft.Json;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public class CleanupVSTSAgent : Microsoft.Build.Utilities.Task
+    {
+        [Required]
+        public string AgentDirectory { get; set; }
+
+        [Required]
+        public double RetentionDays { get; set; }
+
+        public int? Retries { get; set; }
+
+        public int? SleepTimeInMilliseconds { get; set; }
+
+        private static readonly int s_DefaultRetries = 5;
+        private static readonly int s_DefaultSleepTime = 10000;
+
+        public override bool Execute()
+        {
+            if(!Retries.HasValue)
+            {
+                Retries = s_DefaultRetries;
+            }
+            if(!SleepTimeInMilliseconds.HasValue)
+            {
+                SleepTimeInMilliseconds = s_DefaultSleepTime;
+            }
+
+            return CleanupAgentsAsync().Result;
+        }
+        private async System.Threading.Tasks.Task<bool> CleanupAgentsAsync()
+        {
+            bool returnStatus = true;
+            DateTime now = DateTime.Now;
+
+            // Cleanup the agents that the VSTS agent is tracking
+            string[] sourceFolderJsons = Directory.GetFiles(Path.Combine(AgentDirectory, "_work", "SourceRootMapping"), "SourceFolder.json", SearchOption.AllDirectories);
```

Comment:
Looks like you have this in a item group as well as here. I suspect it is probably best to do it here. Mainly because you already made some other assumptions about the directory structure in this task so you would need to change the task no matter what directory changed.